### ISSUE TITLE
[tune] metrics for search algorithm and scheduler?

### ISSUE THREAD

**Issue by JiayuanSternLi** [ray-project/ray#6268]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Ray installed from (source or binary)**:
- **Ray version**:
- **Python version**:
- **Exact command to reproduce**:

### Describe the problem

Take the class-based demo code as an example. Both of the AsyncHyperBandScheduler and the HyperOptSearch have the metric attribute, but how could I select the metrics?? Train loss or test loss, train acc or test acc? 

As far as I know, training metrics must be loss instead of accuracy, but do train_acc and test_acc work better than those in search algorithm or early stopping steps? Do you have any idea or academic result? thanks!

### Source code / logs

```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import argparse
import numpy as np
from tensorflow.keras.layers import Dense, Flatten, Conv2D
from tensorflow.keras import Model
from tensorflow.keras.datasets.fashion_mnist import load_data
from ray.tune.schedulers import AsyncHyperBandScheduler
from ray.tune.suggest.bayesopt import BayesOptSearch
from ray.tune.suggest.hyperopt import HyperOptSearch



from ray import tune

MAX_TRAIN_BATCH = 10
FROTN_END_CALL = False

parser = argparse.ArgumentParser()
parser.add_argument(
    "--smoke-test", action="store_true", help="Finish quickly for testing")
args, _ = parser.parse_known_args()


class MyModel(Model):
    def __init__(self, hiddens=128):
        super(MyModel, self).__init__()
        self.conv1 = Conv2D(32, 3, activation="relu")
        self.flatten = Flatten()
        self.d1 = Dense(hiddens, activation="relu")
        self.d2 = Dense(10, activation="softmax")

    def call(self, x):
        x = self.conv1(x)
        x = self.flatten(x)
        x = self.d1(x)
        return self.d2(x)


class MNISTTrainable(tune.Trainable):
    def _setup(self, config):
        # IMPORTANT: See the above note.
        import tensorflow as tf
        (x_train, y_train), (x_test, y_test) = load_data()
        x_train, x_test = x_train / 255.0, x_test / 255.0

        # Add a channels dimension
        x_train = x_train[..., tf.newaxis]
        x_test = x_test[..., tf.newaxis]
        self.train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
        self.train_ds = self.train_ds.shuffle(10000).batch(
            config.get("batch", 32))
        self.test_ds = tf.data.Dataset.from_tensor_slices((x_test,
                                                           y_test)).batch(32)

        self.model = MyModel(hiddens=config.get("hiddens", 128))
        self.loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
        self.optimizer = tf.keras.optimizers.SGD()

        self.optimizer.learning_rate = config.get("lr", 0.01)
        self.optimizer.momentum = config.get("momentum", 0)
        self.optimizer.nesterov = config.get("nesterov", False)
        self.train_loss = tf.keras.metrics.Mean(name="train_loss")
        self.train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
            name="train_accuracy")

        self.test_loss = tf.keras.metrics.Mean(name="test_loss")
        self.test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
            name="test_accuracy")

        @tf.function
        def train_step(images, labels):
            with tf.GradientTape() as tape:
                predictions = self.model(images)
                print("input size", images.shape)
                loss = self.loss_object(labels, predictions)
            gradients = tape.gradient(loss, self.model.trainable_variables)

            self.optimizer.apply_gradients(
                zip(gradients, self.model.trainable_variables))

            self.train_loss(loss)
            self.train_accuracy(labels, predictions)

        @tf.function
        def test_step(images, labels):
            predictions = self.model(images)
            t_loss = self.loss_object(labels, predictions)

            self.test_loss(t_loss)
            self.test_accuracy(labels, predictions)

        self.tf_train_step = train_step
        self.tf_test_step = test_step

    def _train(self):
        self.train_loss.reset_states()
        self.train_accuracy.reset_states()
        self.test_loss.reset_states()
        self.test_accuracy.reset_states()

        for idx, (images, labels) in enumerate(self.train_ds):
            if idx > MAX_TRAIN_BATCH:  # This is optional and can be removed.
                break
            self.tf_train_step(images, labels)

        for test_images, test_labels in self.test_ds:
            self.tf_test_step(test_images, test_labels)

        # It is important to return tf.Tensors as numpy objects.
        return {
            "epoch": self.iteration,
            "loss": self.train_loss.result().numpy(),
            "accuracy": self.train_accuracy.result().numpy() * 100,
            "test_loss": self.test_loss.result().numpy(),
            "mean_accuracy": self.test_accuracy.result().numpy() * 100
        }


if __name__ == "__main__":
    load_data()  # we download data on the driver to avoid race conditions.
    from hyperopt import hp
    
    space = {
        "lr": hp.uniform("width", 0.001, 0.1),
        "momentum": hp.uniform("height", 0.1, 0.9),
        "hiddens": hp.choice("hiddens", [32, 64, 128]),
        "nesterov": hp.choice("nesterov", [True, False])
    }

    algo = HyperOptSearch(
        space,
        metric="mean_accuracy",
        mode="max"
        )

    scheduler = AsyncHyperBandScheduler(time_attr="training_iteration", 
    									metric="loss", 
    									mode="min", 
    									max_t=100, 
    									grace_period=20)
    analysis = tune.run(
        MNISTTrainable,
        # stop={"training_iteration": 5 if args.smoke_test else 100},
        search_alg = algo,  # comment 1
        verbose=1,
        num_samples = 50,
        scheduler=scheduler,
        # config={
        #         "hiddens": tune.grid_search([32, 64, 128]) # comment
        #        },
        resources_per_trial={'gpu': 8, 'cpu': 40},
        name = "example-experiment"
        )


    print("Best config is", analysis.get_best_config(metric="accuracy"))
    trials = analysis.trials
    print("All trials inform:", trials)
```

**Comment by richardliaw** [comment]
Test accuracy sounds like a good thing to use. Perhaps a better place for this question is stackoverflow?

**Comment by JiayuanSternLi** [comment]
> Test accuracy sounds like a good thing to use. Perhaps a better place for this question is stackoverflow?

Thank you... it sounds like the topic of tune usage should be posted on stackoverflow, while bug report or contribution should be here? Right?

**Comment by richardliaw** [comment]
Yep!
